# scVar:

## Dependence:

## Dependent Filesï¼š

## Each Step:

### Rule cellranger:

Input and dependence:

- fastq files
- reference genome (Hg38)

This step aligns the sequenced reads to the reference genome (Hg38) utilizing multithreading, with the number of threads specified in the configuration file. The resultant matrix and BAM file will be utilized in subsequent analysis steps.

Files generated by this step in the mapping folder:

- **possorted_genome_bam.bam**: This file contains the BAM format output, representing the aligned reads.
- **raw_feature_bc_matrix**: This directory contains all detected barcodes in MEX format.
- **filtered_feature_bc_matrix**: This directory encompasses only the detected cell-associated barcodes in MEX format.

### Rule soupx:

Input and dependence:

- matrix files generated by Cell Ranger
- R script **test_soupx.r**
- shell script **prepare.sh**

This step employs the counts matrix to eliminate cell-free mRNA contamination within droplets, resulting in a filtered single-cell RNA sequencing (scRNA-seq) expression matrix.

Files generated by this step in the mapping folder:

- **soupX_matrix**:  This directory contains the corrected matrix files, reflecting the adjustments made to account for mRNA contamination.

### Rule Seurat:

Input and dependence:

- soupX results located in Cell Ranger result folder
- R script **test.r**
- shell script **seurat_end.sh**

This step proceeds with critical data processing steps:

1. Cell Filtering: The following conditions are applied to retain high-quality cells:
   - Minimum gene count (nFeature_RNA): > 10
   - Maximum gene count (nFeature_RNA): < 15,000
   - Percentage of mitochondrial gene expression: < 10%
   - Percentage of ribosomal gene expression: < 50%
2. Normalization: The count data is normalized to ensure comparability across cells (default scale.factor: 1e4).
3. Feature Selection: This step identifies variable features that contribute significantly to the data variance.
4. Dimensional Reduction: Principal Component Analysis (PCA) is performed to reduce the dimensionality of the dataset.
5. Neighbor Identification: Cells are analyzed to identify neighboring cells based on their expression profiles (default dimensions: 30).
6. Clustering: Utilizing the neighbor relationships, this step clusters the cells into distinct groups.
7. Visualization: The clustering results are visualized through UMAP and t-SNE, providing users with an intuitive representation of the data.
8. Cell Type Annotation: Cell types are annotated using the SingleR package, referencing established transcriptomic datasets for accuracy.

Users have the flexibility to modify parameters as needed by editing the R script **test.r**. This allows for tailored analysis to meet specific research requirements.

Files generated by this step in the seurat folder:

- **barcodes.csv**: Contains the cell barcodes retained after filtering.
- **umap_tsne.png**: Visual representation of clustering results using UMAP and t-SNE.
- **cell_type.png**: Visual representation of clustering results and cell types using UMAP.
- **barcodes_type.csv**: Records all relevant Seurat object information.
  - **barcodes_cell_type.tsv**: Lists the cell type corresponding to each cell.
  - **barcodes_cluster.tsv**: Lists the cluster corresponding to each cell.

### Rule Extract_bam:

Input and dependence:

- **possorted_genome_bam.bam**
- **barcodes_type.csv**
- shell script **pipeline.sh**

This step proceeds with critical data processing steps:

1. Chromosome-Based Splitting: The BAM file is split into subfiles, organized by chromosomes. This allows for more manageable processing and efficient multi-threading.
2. Multi-Threaded Extraction: This process implements a multi-threaded approach to extract reads from each chromosome BAM file. The extraction is guided by barcodes that remain after filtering, utilizing the Python script **multi_extract.py**.
3. Merging Extracted Reads: After extraction, all retrieved reads from the chromosome subfiles are merged into a cohesive final BAM file. This final output contains only the reads associated with the specified barcodes, ready for subsequent analysis.

Files generated by this step in the mapping folder:

- **extracted.bam**: Bam file contains all reads of filtered cells.

### Rule SNV_calling:

Input and dependence:

- **extracted.bam**
- shell script **pipeline.sh**

This step involves a series of critical data processing procedures to accurately identify mutations in the dataset. The following sub-steps are executed:

1. Pre-processing: This stage includes essential tasks such as sorting the BAM files, building their indices, and marking duplicates. These steps ensure that the data is organized and clean for subsequent analysis.
2. SplitNCigarReads: In this step, reads containing "N" characters in their cigar strings are split. This action allows for more precise variant calling by handling reads with unexpected reference gaps appropriately.
3. BaseRecalibrator and ApplyBQSR: This process recalibrates the base quality scores to enhance the accuracy of variant detection.
4. HaplotypeCaller: This final stage identifies SNPs and generates a Variant Call Format (VCF) file. Notably, to maximize the identification of SNPs in 10X data, the ploidy is set to 20, reflecting the high throughput of the data.

Files generated by this step in the mutation folder:

- **variants.vcf**: SNP vcf file.

### Rule genotype:

Input and dependence:

- **variants.vcf**
- **extracted.bam**
- **barcodes.csv**
- shell script **Genotyping.sh**
- shell script **Genotype_Mutations_filter.sh**

**Genotyping.sh** for split tbl file and  genotype each mutation to each cell. limit:

> - Define the minimum number of reads needed to support alt for mutant cells
> - Minimum comparison quality
> - Minimum base mass

**genotype.py** is the primary Python script designed for genotyping. This script utilizes the pysam package to scan BAM files for each mutation, generating following output files.

- ***_All.tsv** This file contains information on all mutations within each chromosome. Each row corresponds to a specific molecule (UB), and it includes:
  - The number of reads supporting the reference allele at the mutation site
  - The number of reads supporting the alternate allele
  - The total number of reads covering that site
- **_counts_CB.tsv** This file records mutations in the cells that cover the mutation site. Each row contains:
  - The number of reads supporting the reference allele at the mutation site in the cell
  - The number of reads supporting the alternate allele in the cell
  - The total number of reads covering that site in the cell

**Genotype_Mutations_filter.sh** for SNV filter with following terms:

> - Minimum number of cells containing mutations
> - VAF

Files generated by this step in the mutation folder:

- **vcf_filtered_final.vcf**: filtered snvs vcf file.

### Rule annotation:

Input and dependence:

- **snp_vaf_final.vcf**
- **conf_test.toml**
- **barcodes_type.csv**
- shell script **annotation.sh**

This step proceeds with critical data processing steps:

1. Annotation: This stage uses Vep and Annovar for functional annotation of mutations and vcfanno to combine information from multiple databases for clinical and cancer information annotation of mutations. Annotation config file is **conf_test.toml**.
2. Normalization: This step uses **normalize_new.py** to output all information specification as a tsv file.
3. Statistic and format conversion: This step uses **normalize_new.py** to output all information specification as a tsv file.

Files generated by this step in the seurat folder:

- **all_final.tsv**: SNP vcf file.

### Rule mutation_signature:
Input and dependence:
- **all_final.tsv**
- shell script **test2.sh**

This step portrays the mutation signature at the level of the whole and the different cell types, respectively. The frequency of mutations in each trinucleotide background was first calculated in each level, and then the similarity was calculated with the latest COSMIC Single Base Substitution (SBS) Signatures.

Files generated by this step in the seurat folder:
* ***_cosmic.csv** : Similarity of mutation signature to COSMIC Signatures overall and in different cell types.
* ***_signature.csv** : The frequency of mutations in each trinucleotide background overall and in different cell types.

### Rule report:
Input and dependence:
- **all_final.tsv**
- shell script **test2.sh**

This step proceeds with critical data processing steps:

1. bam  statistics: This stage counts the average depth and coverage of sequencing on different chromosomes according to the ba m file after cell extraction.
2. top mutations: This stage calculates the mutations that select the top 30 mutant cell counts and highlights the mutant cells for each mutation in the umap and tnse plots in the report/data folder.
3. Statistic: This phase generates a matrix of the number of cells containing mutations per cell type and a matrix of the number of cells containing mutations per cell cluster. At the same time, the number of mutant cells in a particular cell type or cell nest containing the mutation was compared with others for the fisher's test. Finally, a binary matrix of cells and mutations is generated, with 1 indicating the presence of the mutation in the cell and 0 indicating the absence/deficiency of the mutation.
4. Report: This stage visualises the above results in a single html presentation and provides some parameters for top mutation results screening.

Files generated by this step in the report folder:
- **data/*.json** 
- **report.html**